### PR TITLE
Feature/add question set add

### PR DESCRIPTION
**Pull request by Blackolf931**

### CODE REVIEW

**Review by GagarinPavel — APPROVED** [review]

**Review by stas1408123 — APPROVED** [review]

**Review by stas1408123 — APPROVED** [review]

**Review by nantonov — CHANGES_REQUESTED** [review]
Fix smell code from sonar

File: AIS/AIS.API/ViewModels/QuestionSet/QuestionSetAddViewModel.cs [hunk 1]
```diff
@@ -1,7 +1,12 @@
-﻿namespace AIS.API.ViewModels.QuestionSet
+﻿using System.Collections.Generic;
+
+namespace AIS.API.ViewModels.QuestionSet
 {
     public class QuestionSetAddViewModel
     {
         public string Name { get; set; }
+#nullable enable
```
Comment (nantonov) [inline AIS/AIS.API/ViewModels/QuestionSet/QuestionSetAddViewModel.cs:8]: add disable too

File: AIS/AIS.BLL/Interfaces/Services/IQuestionSetService.cs [hunk 2]
```diff
@@ -0,0 +1,11 @@
+﻿using AIS.BLL.Models;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace AIS.BLL.Interfaces.Services
+{
+    public interface IQuestionSetService : IGenericService<QuestionSet>
+    {
+        Task<QuestionSet> Add(QuestionSet entity, CancellationToken ct);
```
Comment (nantonov) [inline AIS/AIS.BLL/Interfaces/Services/IQuestionSetService.cs]: 'IQuestionSetService.Add(QuestionSet, CancellationToken)' hides inherited member 'IGenericService<QuestionSet>.Add(QuestionSet, CancellationToken)'. Use the new keyword if hiding was intended.

File: AIS/AIS.BLL/Models/QuestionSet.cs [hunk 3]
```diff
@@ -8,5 +8,8 @@ public class QuestionSet
         public string Name { get; set; }
         public IEnumerable<QuestionArea> QuestionAreas { get; set; }
         public IEnumerable<Question> Questions { get; set; }
+#nullable enable
```
Comment (nantonov) [inline AIS/AIS.BLL/Models/QuestionSet.cs:11]: Use disable too

File: AIS/AIS.BLL/Services/QuestionSetService.cs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+                new QuestionsQuestionSetsEntity
+            {
+                QuestionSetId = res.Id,
+                QuestionId = x
+            }, ct));
+
+            return _mapper.Map<QuestionSet>(res);
+        }
+
+        private static List<int> ConvertToList(QuestionSet entity)
```
Comment (nantonov) [inline AIS/AIS.BLL/Services/QuestionSetService.cs:46]: Why we cannot use ToList() or smth like that?

File: AIS/AIS.BLL/Services/QuestionSetService.cs [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+            _questionSetRepository = repository;
+            _questionAreasQuestionSetsRepository = questionAreasQuestionSetsRepository;
+            _questionsQuestionSetsRepository = questionsQuestionSetsRepository;
+        }
+        public override async Task<QuestionSet> Add(QuestionSet entity, CancellationToken ct)
+        {
+            var questionAreaIds = ConvertToList(entity);
+            var questionIds = ConvertToList(entity);
+            var res = await _questionSetRepository.Add(_mapper.Map<QuestionSetEntity>(entity), ct);
+            questionAreaIds.ForEach(x => _questionAreasQuestionSetsRepository.Add(
```
Comment (nantonov) [inline AIS/AIS.BLL/Services/QuestionSetService.cs:31]: Just create an entity and save it once

**Review by nantonov — CHANGES_REQUESTED** [review]

**Review by hromau — APPROVED** [review]

File: AIS/AIS.API/ViewModels/QuestionSet/QuestionSetAddViewModel.cs (hunk 1 above)
Comment (Blackolf931) [reply AIS/AIS.API/ViewModels/QuestionSet/QuestionSetAddViewModel.cs:9]: fix

File: AIS/AIS.BLL/Interfaces/Services/IQuestionSetService.cs (hunk 2 above)
Comment (Blackolf931) [reply AIS/AIS.BLL/Interfaces/Services/IQuestionSetService.cs]: fix

File: AIS/AIS.BLL/Models/QuestionSet.cs (hunk 3 above)
Comment (Blackolf931) [reply AIS/AIS.BLL/Models/QuestionSet.cs:12]: fix

File: AIS/AIS.BLL/Services/QuestionSetService.cs (hunk 5 above)
Comment (Blackolf931) [reply AIS/AIS.BLL/Services/QuestionSetService.cs]: fix

File: AIS/AIS.BLL/Services/QuestionSetService.cs (hunk 4 above)
Comment (Blackolf931) [reply AIS/AIS.BLL/Services/QuestionSetService.cs]: fix

File: AIS/AIS.BLL/Services/QuestionSetService.cs [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+        public override async Task<QuestionSet> Add(QuestionSet entity, CancellationToken ct)
+        {
+            var questionAreaIds = entity.QuestionAreaIds?.ToList();
+            var questionIds = entity.QuestionIds?.ToList();
+            var res = await _questionSetRepository.Add(_mapper.Map<QuestionSetEntity>(entity), ct);
+
+            if (questionIds is not null)
+            {
+                var questionsList = questionIds.Select(id => new QuestionsQuestionSetsEntity { QuestionId = id, QuestionSetId = res.Id }).ToList();
+                await _questionsQuestionSetsRepository.AddRange(questionsList.ToList(), ct);
```
Comment (nantonov) [inline AIS/AIS.BLL/Services/QuestionSetService.cs:36]: ToList is redundant

**Review by nantonov — CHANGES_REQUESTED** [review]

File: AIS/AIS.BLL/Services/QuestionSetService.cs (hunk 6 above)
Comment (Blackolf931) [reply AIS/AIS.BLL/Services/QuestionSetService.cs]: fix

**Review by nantonov — APPROVED** [review]

**Pull request merged**